File: .github/fixtures/profiles/combined-resources/scripts/normalize.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

begin
  unless ARGV.length == 2
    warn "usage: ruby scripts/normalize.rb INPUT OUTPUT"
    exit 2
  end

  input_path, output_path = ARGV
  same_path = File.expand_path(input_path) == File.expand_path(output_path)
  same_file = File.exist?(input_path) && File.exist?(output_path) &&
              File.identical?(input_path, output_path)
  if same_path || same_file
    warn "input and output must refer to different files"
    exit 2
  end

  text = File.binread(input_path).force_encoding(Encoding::UTF_8)
  unless text.valid_encoding?
    warn "invalid UTF-8 input"
    exit 3
  end

  normalized = text.gsub("\r\n", "\n").gsub("\r", "\n")
  normalized = normalized.lines.map { |line| line.sub(/[\t ]+(?=\n?\z)/, "") }.join
  normalized = "#{normalized.sub(/\n*\z/, "")}\n"
  File.binwrite(output_path, normalized)
  puts output_path
rescue SystemCallError => e
  warn e.message
  exit 1
end

File: .github/fixtures/profiles/script-assisted-runtime/scripts/normalize.py
#!/usr/bin/env python3
"""Normalize one UTF-8 file using only the Python standard library."""

from __future__ import annotations

import os
import re
import sys
from pathlib import Path


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: python scripts/normalize.py INPUT OUTPUT", file=sys.stderr)
        return 2

    input_text, output_text = argv
    input_path = Path(input_text)
    output_path = Path(output_text)

    try:
        if output_path.exists() and os.path.samefile(input_path, output_path):
            print("input and output must refer to different files", file=sys.stderr)
            return 2

        raw = input_path.read_bytes()
        try:
            text = raw.decode("utf-8")
        except UnicodeDecodeError:
            print("invalid UTF-8 input", file=sys.stderr)
            return 3

        normalized = text.replace("\r\n", "\n").replace("\r", "\n")
        normalized = re.sub(r"[\t ]+(?=\n|$)", "", normalized)
        normalized = normalized.rstrip("\n") + "\n"
        output_path.write_bytes(normalized.encode("utf-8"))
        print(output_text)
        return 0
    except OSError as exc:
        print(str(exc), file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
